### PR TITLE
simple null checks

### PR DESCRIPTION
**Pull request by dkudernatsch**
Checks if there is a CanBeNullAttribute
Checks if it directly returns null keyword
Checks if it directly returns method that can return null

### CODE REVIEW

File: ConsoleApplication/MethodNullAnnotator.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
+﻿using System;
```
Comment (iam-mholle) [inline ConsoleApplication/MethodNullAnnotator.cs:1]: Missing license header

File: ConsoleApplication/MethodNullAnnotator.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace NullableTypeSetter.ConsoleApplication
+{
+ public class MethodNullAnnotator : CSharpSyntaxRewriter
```
Comment (iam-mholle) [inline ConsoleApplication/MethodNullAnnotator.cs:9]: Wrong indentation

File: ConsoleApplication/MethodNullAnnotator.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      if (node.Body == null)
+        return node;
+      if (node.Body?.Statements.Count == 0)
+        return node;
+
+      var retrunsNull = HasCanBeNullAttribute (node)
+                        || ReturnsNull (node, _semanticModel);
+
+      if (!retrunsNull)
+        return node;
```
Comment (iam-mholle) [inline ConsoleApplication/MethodNullAnnotator.cs:31]:
```suggestion
      var returnsNull = HasCanBeNullAttribute (node)
                        || ReturnsNull (node, _semanticModel);

      if (!returnsNull)
        return node;
```

File: ConsoleApplication/MethodNullAnnotator.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace NullableTypeSetter.ConsoleApplication
```
Comment (iam-mholle) [inline ConsoleApplication/MethodNullAnnotator.cs:7]: `NullableTypeSetter.X` should be `NullableReferenceTypesRewriter.X`

File: ConsoleApplication/MethodNullAnnotator.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+      return returnStatements.Any (
+          stmt => stmt is ReturnStatementSyntax returnStatement
+                  && IsDefinitlyNull (returnStatement.Expression!, semanticModel));
+    }
+
+    private static bool HasCanBeNullAttribute (MethodDeclarationSyntax node)
+    {
+      return node.AttributeLists.SelectMany (list => list.Attributes)
+          .Any (attr => attr.Name.ToString().EndsWith ("CanBeNull"));
```
Comment (iam-mholle) [inline ConsoleApplication/MethodNullAnnotator.cs:52]: Does roslyn expose `CanBeNullAttribute` as `CanBeNull`? Automated tests would really help.

File: ConsoleApplication/MethodNullAnnotator.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    public static bool IsDefinitlyNull (ExpressionSyntax expression, SemanticModel semanticModel)
+    {
+      var typeInfo = semanticModel.GetTypeInfo (expression);
+
+      if (typeInfo.Nullability.FlowState != NullableFlowState.None
+          || typeInfo.Nullability.Annotation == NullableAnnotation.Annotated)
+      {
+        Console.WriteLine (expression);
```
Comment (iam-mholle) [inline ConsoleApplication/MethodNullAnnotator.cs:68]: Why are we writing to console here?

File: ConsoleApplication/MethodNullAnnotator.cs (hunk 2 above)
Comment (iam-mholle) [inline ConsoleApplication/MethodNullAnnotator.cs:9]: ~Consider ordering the methods by accessibility.~
Multiple methods can be made private

File: ConsoleApplication/Program.cs [hunk 7]
```diff
@@ -1,11 +1,89 @@
 ﻿using System;
+using System.IO;
```
Comment (iam-mholle) [inline ConsoleApplication/Program.cs:2]: Missing license header

File: ConsoleApplication/Program.cs [hunk 8]
```diff
@@ -1,11 +1,89 @@
...
+    }
+
+    public static Task<Solution> LoadSolutionSpace (string solutionPath)
+    {
+      var msBuild = MSBuildLocator.QueryVisualStudioInstances().First();
+      MSBuildLocator.RegisterInstance (msBuild);
+      using var workspace = MSBuildWorkspace.Create();
+      workspace.WorkspaceFailed += (sender, args) => { Console.WriteLine (args.Diagnostic); };
+
+      return workspace.OpenSolutionAsync (@"C:\Development\Remotion-git\Remotion.sln");
```
Comment (iam-mholle) [inline ConsoleApplication/Program.cs:86]: This does not seem right 😉 
```suggestion
      return workspace.OpenSolutionAsync (solutionPath);
```

File: ConsoleApplication/Program.cs [hunk 9]
```diff
@@ -1,11 +1,89 @@
...
+      var compilationOptions = project.CompilationOptions as CSharpCompilationOptions;
+
+      compilationOptions = compilationOptions?.WithNullableContextOptions (NullableContextOptions.Enable);
+      if (compilationOptions != null)
+        project = project.WithCompilationOptions (compilationOptions);
+
+      return project;
+    }
+
+    public static Task<Solution> LoadSolutionSpace (string solutionPath)
```
Comment (iam-mholle) [inline ConsoleApplication/Program.cs:79]: Why is this `public`?

**Review by iam-mholle — CHANGES_REQUESTED** [review]
Please add tests.

**Comment by iam-mholle** [comment]
@dkudernatsch Is this one obsolete with https://github.com/rubicon-oss/NullableReferenceTypesRewriter/pull/4?

**Comment by dkudernatsch** [comment]
> 
> 
> @dkudernatsch Is this one obsolete with #4?

yes this was a draft pr anyway and i added lots of new stuff.
The requested changes should be all done however

**Comment by iam-mholle** [comment]
Closed in favor of #4.

**Pull request closed without merging**